### PR TITLE
some options +IP cache + traceroute AS path renderer

### PR DESCRIPTION
**Pull request by pierky**
-dry-run for 'measure' cmd + --ids-only for 'probes'.
New renderer: for each traceroute result show the last `self.RADIUS` ASNs.

Example: https://twitter.com/pierky/status/642295296651714560

```
$ ripe-atlas report 2402742 --renderer traceroute_aspath

RIPE Atlas Report for Measurement #2402742
===================================================

Traceroute measurement to 167.114.29.155

For each traceroute path toward the target, the last 2 ASNs will be shown

Probe #110  :   AS1299  AS16276, completed
Probe #11562:   AS1299  AS16276, completed
Probe #1171 :   AS1299  AS16276, completed
Probe #17651:  AS12714  AS16276, completed
Probe #20001:  AS31133  AS16276, completed
Probe #20242:   AS1299  AS16276, completed
Probe #21542:   AS1299  AS16276, completed
Probe #21566:  AS31133  AS16276, completed
Probe #2859 :   AS1299  AS16276, completed
Probe #3757 :  AS12714  AS16276, completed
Probe #3897 :  AS31133  AS16276, completed
Probe #462  :  AS31133  AS16276, completed
Probe #4706 :   AS1299  AS16276, completed
Probe #4952 :   AS1299  AS16276, completed
Probe #52   :   AS1299  AS16276, completed
Probe #607  :  AS31133  AS16276, completed

Number of probes for each AS path:

   AS12714  AS16276: 2 probes, 2 completed
   AS31133  AS16276: 5 probes, 5 completed
    AS1299  AS16276: 9 probes, 9 completed
```

```
$ ripe-atlas report 2402741 --renderer traceroute_aspath

RIPE Atlas Report for Measurement #2402741
===================================================

Traceroute measurement to 167.114.29.155

For each traceroute path toward the target, the last 2 ASNs will be shown

Probe #110  :   AS3549   AS1299, NOT completed
Probe #11562:   AS2116   AS1299, NOT completed
Probe #1171 :  AS33080   AS1299, NOT completed
Probe #17651:  AS12714  AS31133, NOT completed
Probe #20001:  AS48481  AS31133, NOT completed
Probe #20242:   AS2116   AS1299, NOT completed
Probe #21542: AS198930   AS1299, NOT completed
Probe #21566:   AS8439  AS31133, NOT completed
Probe #2859 :   AS2116   AS1299, NOT completed
Probe #3757 :  AS12714  AS31133, NOT completed
Probe #3897 : AS196695  AS31133, NOT completed
Probe #462  :  AS49534  AS31133, NOT completed
Probe #4706 :  AS14051   AS1299, NOT completed
Probe #4952 : AS197248   AS1299, NOT completed
Probe #52   :  AS30950   AS1299, NOT completed
Probe #607  :  AS52041  AS31133, NOT completed

Number of probes for each AS path:

   AS33080   AS1299: 1 probes, 0 completed
  AS198930   AS1299: 1 probes, 0 completed
   AS12714  AS31133: 2 probes, 0 completed
  AS197248   AS1299: 1 probes, 0 completed
    AS2116   AS1299: 3 probes, 0 completed
   AS52041  AS31133: 1 probes, 0 completed
   AS48481  AS31133: 1 probes, 0 completed
  AS196695  AS31133: 1 probes, 0 completed
    AS3549   AS1299: 1 probes, 0 completed
   AS30950   AS1299: 1 probes, 0 completed
   AS49534  AS31133: 1 probes, 0 completed
    AS8439  AS31133: 1 probes, 0 completed
   AS14051   AS1299: 1 probes, 0 completed
```

### CODE REVIEW

File: ripe/atlas/tools/ipcache.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import time
```
Comment (danielquinn) [inline ripe/atlas/tools/ipcache.py]: I'm not sure about this.  On the one hand, yes, we definitely need a clean
interface that abstracts away the RIPEStat call and caches the output, but I'm
not sure that I like the way this is currently setup: it's effectively a
function wrapped in an object, placed in a variable that then must be treated
like an object to call the method -- it feels convoluted.

Instead, an interface that worked like this would be much cleaner/easier to work
with:

```python
# ip.py

from .cache import cache

class Ip(object):

    def __init__(self, address):
        self.address = address
        self.asn = None
        self.holder = None
        ...
        self._populate()

    def self._populate(self)
        """
        Querys/caches all of the relevant data, populating self.asn, 
        self.holder, etc
        """

        # Pretty much all of your existing code here, but populating self,
        # rather than returning anything


# somewhere_else.py

from ip import Ip

ip = Ip.(address)
print(ip.asn)

ips = Ip.generate(address1, address2 address3)
print(ips[0].asn)
```

Does any of this make sense?  It's just that as it is, it's not really a cache
so much as a model with some built-in caching, and while I like this model, I'd
like it to conform to the so-called "pythonic" style.

Let me know what you think.

File: ripe/atlas/tools/renderers/traceroute_aspath.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        for hop in result.hops:
+            for packet in hop.packets:
+                if packet.origin:
+                    ip_hops.append(packet.origin)
+                    break
+
+        asns = []
+
+        # starting from the last hop's IP, get up to <RADIUS> ASNs
+        for ip in ip_hops[::-1]:
```
Comment (danielquinn) [inline ripe/atlas/tools/renderers/traceroute_aspath.py]: See my comments on the IpCache code you've already got, but this might be better represented as

```python
for address in reversed(ip_hops):  # Use of [::-1] is discouraged
    ip = Ip(address)  # The new Ip object from your other class
    ...
```

File: ripe/atlas/tools/renderers/traceroute_aspath.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+        # Number of different ASs starting from the end of the traceroute path.
+        #
+        #TODO: if a method to pass options to renderers will be defined
+        # this could be set by user input.
+        self.RADIUS = 2
+        
+    @staticmethod
+    def _get_asns_for_output(asns,radius):
+        asns_with_padding = [""]*radius + asns
```
Comment (danielquinn) [inline ripe/atlas/tools/renderers/traceroute_aspath.py]: Nitpick.  pep8 requires spaces on either side of that `*`.

Don't worry, I think this is my last comment :-)

**Comment by pierky** [comment]
You're right, I'll change that and push the changes tonight or tomorrow.

**Comment by danielquinn** [comment]
Thanks!  It's awesome to have someone like you working with us on this.

**Comment by astrikos** [comment]
Hi @pierky, sorry the late response on that. We were busy with some other stuff the last days, so we couldn't really focus on that. Good thing is that I have a version of the new aggregation pushed and probes command is using it now. I will try next to use new aggregator for the report command and migrate your other pull request where you had another type of aggregator(unless you want to rewrite it using the new style aggregators to familiarise with the code).
The new changes in  the probes command unfortunately conflict with your branch now, so the best would be to discard the changes you have (I have already implemented the ids-only option) and then I can go ahead and merge the rest of your stuff here. BTW :+1:  for the tests. We are using nose, so if you run "nosetests tests/" from the main dir you should be able to run all of them.

**Comment by pierky** [comment]
Hi @astrikos, thanks for the feedback.
I removed the `probes` command stuff and merged my new ideas into your current master.
I close the current PR and open a new one; then I'll wait to see the new integration of aggregators.

Just a (stupid) remark: in my original mind the `--ids-only` was intended to build a list of comma-separated IDs that could be used to feed the `measure`'s `--from-probes` argument, in order to build a complex selection of probes that the standard arguments don't allow (by joining multiple `--ids-only` queries executed with different filters). So I decided to print all of them in a single line: with the current implementation a further step is needed to comma-separate them :wink:.

**Comment by astrikos** [comment]
My initial thought was that everyone could use something like tr "\n" "," but yeah you have a point it's easier with a comma. I will change it with the coming commits. Thanks!

**Comment by astrikos** [comment]
@pierky I just noticed that the majority of these tests are querying stat.ripe.net, and they break if you are offline, which is not very optimal. As an enhancement (maybe when there is more time :)) you could decouple part of code that it's doing the actually http request and have it in a small function. This way you can write a single test for this function (which will be the only one to break in offline cases, or even better mock urllib and never go to stat) and for the rest you can easily mock this function with whatever return value stat would give. 
Also, this way we keep running time of tests smaller.
```bash
.................
tests.test_ipdetails.TestIPDetails.test_fakecache_diffprefix: 0.5293s
tests.test_ipdetails.TestIPDetails.test_fakecache_notannounced: 0.2742s
tests.test_ipdetails.TestIPDetails.test_nocache: 0.2579s
tests.test_ipdetails.TestIPDetails.test_fakecache_sameip: 0.2466s
tests.test_ipdetails.TestIPDetails.test_fakecache_sameprefix: 0.2453s
tests.TestProbesCommand.test_arg_with_wrong_type: 0.0061s
tests.TestProbesCommand.test_arg_with_no_value: 0.0022s
....
tests.aggregators.test_aggregators.TestAggregators.test_mixed_aggregation: 0.0012s
tests.test_ipdetails.TestIPDetails.test_loopback4: 0.0012s
tests.test_ipdetails.TestIPDetails.test_loopback6: 0.0012s
...
Ran 60 tests in 1.617s
```
Notice that whatever goes to stat real time it's more expensive.

**Comment by astrikos** [comment]
Also since we are using unittest lib you could use [setUpClass](https://docs.python.org/2/library/unittest.html#setupclass-and-teardownclass) instead of mock_setUp. This way you don't have to call it each time for every test, setUpClass is called automatically before each test for you.
Sorry, for these late comments but I just had a case where tests where broken because I was offline and took some more time to see this code more carefully.

**Comment by pierky** [comment]
In the beginning I also thought to use cached results to improve response times, but then I opted for the real time data in order to check both current RIPEstat's output format and reachability. But you are right, I can add a single test to test both once, then relay on cached results.
Any suggestion where to store cached JSON blobs? External files, within the code as it is now, ...

For the setUpClass, I'll work on it later/next days.

**Comment by danielquinn** [comment]
Typical unit test practise is to not depend on external sources that may change or be up/down/whatever.  The thinking is that you're building software that talks to known externals, and if those externals change, that's not the responsibility of your project.

If you're really concerned that the external process might change, then you can add an explicit test for that and then make it easy for people to skip it in most cases (tests should be fast).

To test code that depends on these external services, [mock](https://pypi.python.org/pypi/mock) is your friend.  This way, you're left testing only the unit in question and not (potentially):

* Your connectivity
* Your HTTP stack (urllib/requests/whatever)
* Your caching layer
* The service you're communicating with
* The data provided by that service

Instead, you're essentially saying *"I expect that if `my_method` receives `this input`, it will return `this output`"* -- which is all that you should expect from a unit test.

So, TL;DR and in answer to your question, I don't recommend any type of caching for unit tests, because unit tests need to test units and not external services.

Also, it doesn't hurt that we maintain those services, so we can be prepared if things are about to change ;-)

**Comment by astrikos** [comment]
I think what @pierky meant was where to have the mocked responses from RIPEstat. Maybe cache as word was a bit misleading. He can confirm himself if that what it is, but I think the change he made are towards the correct direction. Now, we have only one function that can break if stat breaks, if we get rid of this as well at some point we are perfect :)
